amundsen_application/static/js/components/TableDetail/ColumnListItem/ColumnType/parser.ts

```diff
@@ -0,0 +1,126 @@
+  ')': '(',
+  '>': '<',
+  ']': '[',
+};
+const SEPARATOR_DELIMETER = ',';
+
+/*
+ * Iterates through the columnType string and recursively creates a NestedType
+ */
+function parseNestedTypeHelper(
+  columnType: string,
+  startIndex: number = 0,
+  currentIndex: number = 0
+): { nextStartIndex: number; results: ParsedType[] } {
+  const children: ParsedType[] = [];
+
+  while (currentIndex < columnType.length) {
+    const currentChar = columnType.charAt(currentIndex);
+
+    if (currentChar === SEPARATOR_DELIMETER) {
+      /* Case 1: End of non-nested item */
+      children.push(columnType.substring(startIndex, currentIndex + 1));
+      startIndex = currentIndex + 1;
+      currentIndex = startIndex;
+    } else if (currentChar in CLOSE_DELIMETERS) {
+      /* Case 2: End of a nested item  */
+      if (startIndex !== currentIndex) {
+        children.push(columnType.substring(startIndex, currentIndex));
+      }
+      return {
+        nextStartIndex: currentIndex + 1,
+        results: children,
+      };
+    } else if (currentChar in OPEN_DELIMETERS) {
+      /* Case 3: Beginning of a nested item */
+      const parsedResults = parseNestedTypeHelper(
+        columnType,
+        currentIndex + 1,
+        currentIndex + 1
+      );
+      let isLast: boolean = true;
+      let { nextStartIndex } = parsedResults;
+
+      if (columnType.charAt(nextStartIndex) === SEPARATOR_DELIMETER) {
+        isLast = false;
+        nextStartIndex++;
+      }
+
+      children.push({
+        head: columnType.substring(startIndex, currentIndex + 1),
+        tail: `${OPEN_DELIMETERS[currentChar]}${
+          isLast ? '' : SEPARATOR_DELIMETER
+        }`,
+        children: parsedResults.results,
+      });
+
+      startIndex = nextStartIndex;
+      currentIndex = startIndex;
+    } else {
+      currentIndex++;
+    }
+  }
+
+  return {
+    nextStartIndex: currentIndex + 1,
+    results: children,
+  };
+}
+
+/*
+ * Returns whether or not a columnType string represents a complexc type for the given database
+ */
+function isNestedType(columnType: string, databaseId: string): boolean {
+  const supportedTypes = SUPPORTED_TYPES[databaseId];
+  let isNestedType = false;
+  supportedTypes.forEach((supportedType) => {
+    if (columnType.startsWith(supportedType) && columnType !== supportedType) {
+      isNestedType = true;
+    }
+  });
+  return isNestedType;
+}
+
+/**
+ * Returns a NestedType object for supported complex types, else returns null
+ */
+export function parseNestedType(
+  columnType: string,
+  databaseId: string
+): NestedType | null {
+  if (isNestedType(columnType, databaseId)) {
+    return parseNestedTypeHelper(columnType).results[0] as NestedType;
+  }
+  return null;
+}
+
+/*
+ * Returns the truncated string representation for a NestedType
+ */
+export function getTruncatedText(nestedType: NestedType): string {
+  const { head, tail } = nestedType;
+  return `${head}...${tail.replace(SEPARATOR_DELIMETER, '')}`;
+}
```

Comment:
It would be really useful to have tests for this kind of logic, as it is not trivial.

Reply:
I plan on it but I'm mostly a "write tests when I'm sure of the implementation" person, but I'll add tests for these smaller functions because its really the `parseNestedTypeHelper` that I want manually test on a greater sample size. (i.e. trying to find a very complex case of each type for each database to check any assumptions I might have made, and then use those kinds of cases for unit tests)